File: 21-customer-address/Customer.java
/*
 * Ngôn ngữ: Java
 * Công dụng: Customer/Address minh họa NULL OBJECT PATTERN — thay vì biểu
 * diễn "khách chưa có địa chỉ" bằng null rồi rải if (address == null) khắp
 * nơi, ta tạo một Address ĐẶC BIỆT (UNKNOWN) đại diện cho sự vắng mặt đó.
 * UNKNOWN vẫn là một Address thật — vẫn gọi được getDisplayText() bình
 * thường, tự nó biết trả lời "chưa có địa chỉ" thay vì bắt nơi gọi phải
 * kiểm tra null trước mỗi lần dùng.
 * Tại sao cần học: bài 13 (Temperature) dùng static factory để tạo object
 * đại diện cho một GIÁ TRỊ CÓ THẬT. Bài này dùng đúng kỹ thuật đó nhưng cho
 * mục đích ngược lại — tạo object đại diện cho SỰ VẮNG MẶT của giá trị.
 */
class Address { // đại diện một địa chỉ giao hàng — hoặc "chưa có địa chỉ nào"
    public static final Address UNKNOWN = new Address("", "", true); // hằng số dùng chung — MỘT object duy nhất đại diện "chưa có địa chỉ"

    private final String street; // tên đường — rỗng nếu là UNKNOWN
    private final String city;   // thành phố — rỗng nếu là UNKNOWN
    private final boolean unknown; // cờ đánh dấu đây có phải Address đặc biệt (UNKNOWN) hay không

    private Address(String street, String city, boolean unknown) { // constructor riêng tư — chỉ dùng nội bộ (bởi UNKNOWN và of())
        this.street = street;
        this.city = city;
        this.unknown = unknown;
    }

    public static Address of(String street, String city) { // static factory — tạo địa chỉ CỤ THỂ, có thật
        if (street == null || street.isEmpty()) throw new IllegalArgumentException("Tên đường không được rỗng"); // validate ở biên
        if (city == null || city.isEmpty()) throw new IllegalArgumentException("Thành phố không được rỗng"); // validate ở biên
        return new Address(street, city, false); // false: không phải UNKNOWN
    }

    public boolean isUnknown() { // kiểm tra: đây có phải Address "chưa có gì" hay không
        return unknown;
    }

    public String getDisplayText() { // TỰ BIẾT hiển thị đúng cách — nơi gọi không cần kiểm tra null/unknown trước
        if (unknown) return "Chưa có địa chỉ giao hàng"; // hành vi an toàn khi không có dữ liệu thật
        return street + ", " + city; // hành vi bình thường khi có dữ liệu thật
    }
}

public class Customer { // khai báo class — khuôn mẫu để tạo ra các object Customer

    private final String name;    // tên khách hàng — riêng cho từng object
    private final Address address; // KHÔNG BAO GIỜ null — luôn là Address.UNKNOWN hoặc một Address cụ thể

    public Customer(String name) { // constructor — khách CHƯA có địa chỉ, tự gán Address.UNKNOWN thay vì null
        this(name, Address.UNKNOWN); // ủy quyền sang constructor đầy đủ (giống this(...) ở bài 07)
    }

    public Customer(String name, Address address) { // constructor — khách có địa chỉ cụ thể ngay từ đầu
        if (name == null || name.isEmpty()) throw new IllegalArgumentException("Tên khách hàng không được rỗng"); // validate ở biên
        this.name = name;
        this.address = address; // luôn có giá trị hợp lệ — không cần kiểm tra null ở đây hay bất kỳ đâu khác
    }

    public boolean hasAddress() { // có địa chỉ THẬT hay không — hỏi thẳng Address thay vì so sánh với null
        return !address.isUnknown();
    }

    public String getShippingLabel() { // nhãn giao hàng — KHÔNG cần if (address == null): address.getDisplayText() tự lo phần đó
        return name + ": " + address.getDisplayText();
    }

    // ---- Self-check ----
    private static void check(boolean ok, String msg) { // hàm phụ trợ: nếu "ok" sai thì báo lỗi kèm thông điệp "msg"
        if (!ok) throw new AssertionError(msg); // ném lỗi để dừng ngay và chỉ rõ chỗ sai
    }

    public static void main(String[] args) { // điểm bắt đầu chạy chương trình khi gõ "java Customer"
        Customer guest = new Customer("An"); // chưa cung cấp địa chỉ — tự động là Address.UNKNOWN, không phải null
        check(!guest.hasAddress(), "khách chưa cung cấp địa chỉ thì hasAddress() phải là false");
        check(guest.getShippingLabel().equals("An: Chưa có địa chỉ giao hàng"), "nhãn giao hàng khi chưa có địa chỉ phải đúng thông báo mặc định");

        Address home = Address.of("123 Lê Lợi", "Đà Nẵng"); // địa chỉ cụ thể, có thật
        Customer binh = new Customer("Binh", home);
        check(binh.hasAddress(), "khách có địa chỉ cụ thể thì hasAddress() phải là true");
        check(binh.getShippingLabel().equals("Binh: 123 Lê Lợi, Đà Nẵng"), "nhãn giao hàng khi có địa chỉ phải in đúng street, city");

        check(Address.UNKNOWN.isUnknown(), "Address.UNKNOWN phải tự nhận là unknown");
        check(!home.isUnknown(), "Address cụ thể không được coi là unknown");

        boolean blockedStreet = false; // cờ đánh dấu: đường rỗng có bị chặn không
        try { Address.of("", "Hà Nội"); } catch (IllegalArgumentException e) { blockedStreet = true; }
        check(blockedStreet, "tên đường rỗng phải bị chặn khi tạo Address cụ thể");

        boolean blockedName = false; // cờ đánh dấu: tên khách hàng rỗng có bị chặn không
        try { new Customer(""); } catch (IllegalArgumentException e) { blockedName = true; }
        check(blockedName, "tên khách hàng rỗng phải bị chặn");

        System.out.println("OK"); // in "OK" nếu mọi check ở trên đều đúng
    }
}

File: 21-customer-address/Customer.cpp
/*
 * Ngôn ngữ: C++
 * Công dụng: Customer/Address minh họa NULL OBJECT PATTERN — thay vì biểu
 * diễn "khách chưa có địa chỉ" bằng một con trỏ null rồi kiểm tra khắp nơi,
 * ta tạo một Address ĐẶC BIỆT (UNKNOWN) đại diện cho sự vắng mặt đó. UNKNOWN
 * vẫn là một Address thật — vẫn gọi được getDisplayText() bình thường, tự nó
 * biết trả lời "chưa có địa chỉ" thay vì bắt nơi gọi phải kiểm tra null.
 * Tại sao cần học: bài 13 (Temperature) dùng static factory để tạo object
 * đại diện cho một GIÁ TRỊ CÓ THẬT. Bài này dùng đúng kỹ thuật đó nhưng cho
 * mục đích ngược lại — tạo object đại diện cho SỰ VẮNG MẶT của giá trị, và vì
 * Address là kiểu GIÁ TRỊ (không phải con trỏ), không hề có nguy cơ null.
 */
#include <cstdlib>   // thư viện chuẩn: std::exit để dừng chương trình khi self-check thất bại
#include <iostream>  // thư viện chuẩn: std::cout/std::cerr để in ra màn hình
#include <stdexcept> // thư viện chuẩn: std::invalid_argument để báo lỗi dữ liệu sai
#include <string>    // thư viện chuẩn: std::string cho tên/địa chỉ

class Address { // đại diện một địa chỉ giao hàng — hoặc "chưa có địa chỉ nào"
private:
    std::string street; // tên đường — rỗng nếu là UNKNOWN
    std::string city;   // thành phố — rỗng nếu là UNKNOWN
    bool unknown;        // cờ đánh dấu đây có phải Address đặc biệt (UNKNOWN) hay không

    Address(std::string street, std::string city, bool unknown) // constructor riêng tư — chỉ dùng nội bộ (bởi unknown() và of())
        : street(std::move(street)), city(std::move(city)), unknown(unknown) {}

public:
    static Address unknown_address() { // "hằng số" — tạo Address đặc biệt đại diện "chưa có địa chỉ" (C++ không có static field khởi tạo phức tạp gọn như Java, nên dùng hàm)
        return Address("", "", true);
    }

    static Address of(std::string street, std::string city) { // static factory — tạo địa chỉ CỤ THỂ, có thật
        if (street.empty()) throw std::invalid_argument("Tên đường không được rỗng"); // validate ở biên
        if (city.empty()) throw std::invalid_argument("Thành phố không được rỗng"); // validate ở biên
        return Address(std::move(street), std::move(city), false); // false: không phải unknown
    }

    bool isUnknown() const { // kiểm tra: đây có phải Address "chưa có gì" hay không
        return unknown;
    }

    std::string getDisplayText() const { // TỰ BIẾT hiển thị đúng cách — nơi gọi không cần kiểm tra trước
        if (unknown) return "Chua co dia chi giao hang"; // hành vi an toàn khi không có dữ liệu thật
        return street + ", " + city; // hành vi bình thường khi có dữ liệu thật
    }
};

class Customer { // khai báo class — khuôn mẫu để tạo ra các object Customer
private:
    std::string name;  // tên khách hàng — riêng cho từng object
    Address address;    // KHÔNG BAO GIỜ là con trỏ null — luôn là Address::unknown_address() hoặc một Address cụ thể (kiểu giá trị)

public:
    explicit Customer(std::string name) // constructor — khách CHƯA có địa chỉ, tự gán Address unknown thay vì null
        : Customer(std::move(name), Address::unknown_address()) {} // ủy quyền sang constructor đầy đủ (giống this(...) ở bài 07)

    Customer(std::string name, Address address) // constructor — khách có địa chỉ cụ thể ngay từ đầu
        : address(std::move(address)) {
        if (name.empty()) throw std::invalid_argument("Ten khach hang khong duoc rong"); // validate ở biên
        this->name = std::move(name);
    }

    bool hasAddress() const { // có địa chỉ THẬT hay không — hỏi thẳng Address thay vì so sánh với null
        return !address.isUnknown();
    }

    std::string getShippingLabel() const { // nhãn giao hàng — KHÔNG cần kiểm tra null: address.getDisplayText() tự lo phần đó
        return name + ": " + address.getDisplayText();
    }
};

// ---- Self-check ----
static void check(bool ok, const std::string& msg) { // hàm phụ trợ: nếu "ok" sai thì báo lỗi kèm thông điệp "msg"
    if (!ok) { std::cerr << "FAIL: " << msg << "\n"; std::exit(1); } // in chỗ sai rồi dừng ngay
}

int main() { // điểm bắt đầu chạy chương trình sau khi biên dịch
    Customer guest("An"); // chưa cung cấp địa chỉ — tự động là Address unknown, không phải con trỏ null
    check(!guest.hasAddress(), "khach chua cung cap dia chi thi hasAddress() phai la false");
    check(guest.getShippingLabel() == "An: Chua co dia chi giao hang", "nhan giao hang khi chua co dia chi phai dung thong bao mac dinh");

    Address home = Address::of("123 Le Loi", "Da Nang"); // địa chỉ cụ thể, có thật
    Customer binh("Binh", home);
    check(binh.hasAddress(), "khach co dia chi cu the thi hasAddress() phai la true");
    check(binh.getShippingLabel() == "Binh: 123 Le Loi, Da Nang", "nhan giao hang khi co dia chi phai in dung street, city");

    check(Address::unknown_address().isUnknown(), "Address unknown phai tu nhan la unknown");
    check(!home.isUnknown(), "Address cu the khong duoc coi la unknown");

    bool blockedStreet = false; // cờ đánh dấu: đường rỗng có bị chặn không
    try { Address::of("", "Ha Noi"); } catch (const std::invalid_argument&) { blockedStreet = true; }
    check(blockedStreet, "ten duong rong phai bi chan khi tao Address cu the");

    bool blockedName = false; // cờ đánh dấu: tên khách hàng rỗng có bị chặn không
    try { Customer bad(""); } catch (const std::invalid_argument&) { blockedName = true; }
    check(blockedName, "ten khach hang rong phai bi chan");

    std::cout << "OK\n"; // in "OK" nếu mọi check ở trên đều đúng
    return 0; // báo chương trình kết thúc thành công
}
